app/server/toronto-bids/toronto-bids/src/toronto_bids/documents.clj

```diff
@@ -0,0 +1,98 @@
+(ns toronto-bids.documents
+	(:gen-class)
+	(:require [clojure.string :as str])
+	(:require [clojure.java.jdbc :as jdbc])
+	(:require [clojure.data.json :as json])
+)
+
+(def SEARCH-STRING 
+	(str 
+		"document.id IN (SELECT id FROM" 
+		" (SELECT id, short_description, description, search_text FROM document" 
+		" HAVING MATCH(short_description, description, search_text) AGAINST (? IN BOOLEAN MODE)) AS found)"
+	)
+)
+
+; The argument-string
+; (call_number type division commodity commodity_type posting_date_after posting_date_before closing_date_after closing_date_before buyer search-text) 
+
+(def TEST-LIST 
+	(list "call_number = ?" "type = ?" "division = ?" "commodity = ?" "commodity_type = ?" 
+	"posting_date >= ?" "posting_date <= ?" "closing_date >= ?" "closing_date <= ?" "buyer = ?" SEARCH-STRING)
+)
+
+(def COLUMN-STRING " document.id AS document_id, type, call_number, commodity, commodity_type, short_description, posting_date, closing_date, site_meeting")
+
+(def FROM-STRING
+	(str " FROM document" 	
+		" JOIN type ON document.type_id = type.id"
+		" JOIN division ON document.division_id = division.id"
+		" JOIN commodity_type ON document.commodity_type_id = commodity_type.id"
+		" JOIN commodity ON commodity_type.commodity_id = commodity.id"
+		" JOIN document_buyer ON document.id = document_id"
+		" JOIN buyer ON buyer_id = buyer.id"
+	)
+)
+
+(def HEAD (str "SELECT DISTINCT" COLUMN-STRING FROM-STRING))
+
+(def ORDER-STRING " ORDER BY posting_date")
+
+(def BUYER-SELECT 
+	"SELECT buyer, phone, email, location FROM document_buyer JOIN buyer ON buyer_id = buyer.id JOIN location ON location_id = location.id WHERE document_id = ?;"
+)
+
+(defn where-phrase [test start]
+	(if start (str " WHERE " test) (str " AND " test))
```

Comment:
To reduce branching here, could you conditionally include the 'where' keyword as part of the logic of the 'HEAD' method?

Reply:
Don't think that would work.  If the WHERE is part of HEAD then where-phrase needs to conditionally start with blank.  Could make the HEAD include WHERE TRUE,so everything starts with AND.

Reply:
I have implemented WHERE TRUE in the HEAD.